README.md
# IP to Country

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

IP to Country is a module for detecting country information of IP addresses without any API call.

## Installation

```bash
npm install iptocountry
```

## Usage

### REST API

```bash
npm start

# See http://localhost:12080/search?ip=74.125.225.71
# See http://localhost:12080/search?ip=74.125.225.71,98.138.253.109
```

### Download IP database

```javascript
var ip2co = require('iptocountry');

if(ip2co.dbCSVCheckExp(48)) {
  ip2co.dbGet().then(function() {
    console.log('done!');
  }, function(err) {
    console.log(err);
  });
}
// done!
```

### Search IP addresses

```javascript
var ip2co = require('iptocountry');

var ipSearch = ip2co.ipSearch(['74.125.225.71', '98.138.253.109']);
if(!ipSearch.error) {
  console.log(JSON.stringify(ipSearch, null, 2));
} else {
  console.log(ipSearch.error);
};

/*
{
  "data": {
    "74.125.225.71": {
      "ip": "74.125.225.71",
      "ipNum": 1249763655,
      "registery": "arin",
      "assigned": "1173744000",
      "coCode2": "US",
      "coCode3": "USA",
      "country": "United States",
      "time": "2007-03-13 00:00:00",
      "found": true
    },
    "98.138.253.109": {
      "ip": "98.138.253.109",
      "ipNum": 1653276013,
      "registery": "arin",
      "assigned": "1196985600",
      "coCode2": "US",
      "coCode3": "USA",
      "country": "United States",
      "time": "2007-12-07 00:00:00",
      "found": true
    }
  },
  "warnings": [],
  "timeE": 51
}
*/
```

### Listen HTTP requests

```javascript
var ip2co = require('iptocountry');

ip2co.dbLoad();
ip2co.listenHTTP({hostname: 'localhost', 'port': 12080});

// See http://localhost:12080/search?ip=74.125.225.71
```

For performance test; `ab -n 1000 http://localhost:12080/search?ip=74.125.225.71`

## Notes

* It uses [Webnet77](http://software77.net/geo-ip/) for IP database file. It represents
  the over 4 billion IPV4 numbers as well as the virtually inexhaustible IPV6 range (3.4e+38).
  There is limit for downloads, for more information see [FAQ](http://software77.net/faq.html)

### Implementations

- [x] IPv4
- [ ] IPv6

## License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/iptocountry
[npm-image]: https://badge.fury.io/js/iptocountry.svg

[travis-url]: https://travis-ci.org/devfacet/iptocountry
[travis-image]: https://travis-ci.org/devfacet/iptocountry.svg?branch=master


CHANGELOG.md
# Changelog

## 1.0.2 (2017-03-05)

- Maintenance

## 1.0.1 (2015-03-15)

- Misc. updates

## 1.0.0 (2014-07-04)

- Stable release
- Update utilex module

## 0.2.4 (2014-05-22)

- Update badges
- Misc. updates
- Change JS linter & ignore rules

## 0.2.3 (2014-04-29)

- Add limit for IP addresses (HTTP)
- Add --listen-http option
- Add Dependency Status

## 0.2.2 (2014-04-26)

- Add unit test
- Add npm badge
- Add Build Status
- Add .travis.yml

## 0.2.1 (2014-04-21)

- Misc. updates

## 0.2.0 (2014-04-13)

- Performance improvement
- Add HTTP support
- Use in-memory search instead of file search

## 0.1.2 (2014-03-31)

- Misc. changes

## 0.1.1 (2014-03-31)

- Changes for examples

## 0.1.0 (2014-03-31)

- IPv4 implementation
- Adding ipSearch function
- Adding dbCSVCheckExp function
- Adding dbCSVCheck function
- Adding dbGet function

## 0.0.1 (2014-03-27)

- Alpha release

test/test-all.js
/* jslint node: true */
/* global describe: false, it: false */
'use strict';

var ip2co  = require('../'),
    expect = require('chai').expect;

// Tests

describe('iptocountry', function() {

  describe('dataDirCheck', function() {
    var dataDirCheck = ip2co.dataDirCheck();

    it('should check data directory (' + dataDirCheck + ')', function(done) {
      expect(dataDirCheck).to.be.a('boolean');
      done();
    });
  });

  describe('dbFileCheck', function() {
    var dbFileCheck = ip2co.dbFileCheck();

    it('should check database file (' + dbFileCheck + ')', function(done) {
      expect(dbFileCheck).to.be.a('boolean');
      done();
    });
  });

  describe('dbCSVCheck', function() {
    var dbCSVCheck = ip2co.dbCSVCheck();

    it('should check CSV file (' + dbCSVCheck + ')', function(done) {
      expect(dbCSVCheck).to.be.a('boolean');
      done();
    });
  });

  describe('dbCSVCheckExp', function() {
    var dbCSVCheckExp = ip2co.dbCSVCheckExp(48);

    it('should check expiration (' + dbCSVCheckExp + ')', function(done) {
      expect(dbCSVCheckExp).to.be.a('boolean');
      done();
    });
  });

  describe('dbLoad', function() {
    var dbLoad = ip2co.dbCSVCheck();

    it('should load database (' + dbLoad + ')', function(done) {
      expect(dbLoad).to.be.a('boolean');
      done();
    });
  });

  describe('ipSearch', function() {
    var ipSearch = ip2co.ipSearch(['74.125.225.71', '98.138.253.109']);

    if(ip2co.dbCSVCheck()) {
      it('should search and find IP addresses', function(done) {

        if(ipSearch.error) return done(ipSearch.error);

        var ipToCheck = '74.125.225.71';
        expect(ipSearch).to.be.a('object');
        expect(ipSearch).to.have.property('data');
        expect(ipSearch.data).to.be.a('object');
        expect(ipSearch.data).to.have.property(ipToCheck);
        expect(ipSearch.data[ipToCheck]).to.have.property('ip');
        expect(ipSearch.data[ipToCheck]).to.have.property('ipNum');
        expect(ipSearch.data[ipToCheck]).to.have.property('registery');
        expect(ipSearch.data[ipToCheck]).to.have.property('assigned');
        expect(ipSearch.data[ipToCheck]).to.have.property('coCode2');
        expect(ipSearch.data[ipToCheck]).to.have.property('coCode3');
        expect(ipSearch.data[ipToCheck]).to.have.property('country');
        expect(ipSearch.data[ipToCheck]).to.have.property('time');
        expect(ipSearch.data[ipToCheck]).to.have.property('found', true);
        done();
      });
    } else {
      it('ipSearch() should fail due missing database file (' + ipSearch.error + ')', function(done) {
        if(!ipSearch.error) return done('Should give error!');

        done();
      });
    }
  });
});

test/test-http.js
// Init reqs
/* jslint node: true */
'use strict';

var ip2co  = require('../'),
    utilex = require('utilex')
;

// Init vars
var appArgs   = utilex.tidyArgs(),  // args
    appConfig = {isHeapdump: false} // config
;

// config
if(typeof appArgs['heapdump'] !== 'undefined') appConfig.isHeapdump = true;

// heapdump
if(appConfig.isHeapdump === true) var heapdump = require('heapdump');

// Tests
utilex.tidyLog('test-http.js');

// Loads the database and listen http requests.
function loadAndServe() {
  ip2co.dbLoad();
  ip2co.listenHTTP({hostname: 'localhost', 'port': 12080});
}

if(ip2co.dbCSVCheckExp(48)) {
  ip2co.dbGet().then(function() {
    loadAndServe();
  }, function(err) {
    utilex.tidyLog(err);
  });
} else {
  loadAndServe();
}

// heapdump
if(appConfig.isHeapdump === true) heapdump.writeSnapshot(__dirname + '/dump-' + Date.now() + '.heapsnapshot');

lib/iptocountry.js
/*
 * IP to Country
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var fs     = require('fs'),
    path   = require('path'),
    utilex = require('utilex'),
    zlib   = require('zlib'),
    q      = require('q'),
    http   = require('http'),
    url    = require('url');

// Init the module
module.exports = function() {

  var gzip           = zlib.createGunzip(),
      dataSet        = [],
      dataDirPath    = fs.realpathSync(__dirname + path.sep + '..') + path.sep + 'data',
      dbFileUrl      = 'http://software77.net/geo-ip/?DL=1',
      dbFilePath     = dataDirPath + path.sep + 'IpToCountry.csv.gz',
      dbCSVPath      = dataDirPath + path.sep + 'IpToCountry.csv',
      dataDirCheck,  // checks the data folder - function
      dbFileCheck,   // checks the database file - function
      dbCSVCheck,    // checks the database csv file - function
      dbCSVCheckExp, // checks the database csv file's expiration - function
      dbGet,         // gets the database - function
      dbLoad,        // loads the database - function
      ipSearch,      // search ip addresses - function
      listenHTTP;    // listen http requests - function

  // Checks whether the data directory exists and writable or not.
  dataDirCheck = function dataDirCheck() {
    return utilex.dirIsWritable(dataDirPath);
  };

  // Checks whether the database file exists or not.
  dbFileCheck = function dbFileCheck() {
    return fs.existsSync(dbFilePath);
  };

  // Checks whether the database csv file exists or not.
  dbCSVCheck = function dbCSVCheck() {
    return fs.existsSync(dbCSVPath);
  };

  // Checks whether the database csv file expired or not.
  dbCSVCheckExp = function dbCSVCheckExp(hour) {

    if(!dbCSVCheck()) return true;
    if(isNaN(parseInt(('' + hour), 10))) return true;

    var ss      = fs.statSync(dbCSVPath),
        modTime = (ss && ss.mtime && ss.mtime.getTime) ? ss.mtime.getTime() : 0,
        curTime = (new Date()).getTime(),
        difHrs  = (modTime) ? Math.floor((curTime-modTime)/(60*60*1000)) : 0,
        expHr   = parseInt(('' + hour), 10);

    if(expHr <= difHrs) return true;

    return false;
  };

  // Gets the database.
  dbGet = function dbGet() {

    var deferred = q.defer();

    // Get the file
    utilex.download(dbFileUrl, dbFilePath, function(err, fp) {
      if(err) deferred.reject(err);

      // Save the file
      var rs = fs.createReadStream(fp),
          ws = fs.createWriteStream(dbCSVPath),
          ps = rs.pipe(gzip).pipe(ws)
      ;

      ps.on('finish', function() {
        deferred.resolve(null);
      });
      ps.on('error', function(err) {
        deferred.reject(err);
      });
    });

    return deferred.promise;
  };

  dbLoad = function dbLoad(reload) {

    // Check the database
    if(!dbCSVCheck()) {
      return false;
    } else if(dataSet.length > 0 && reload !== true) {
      return true;
    }

    // Read the file
    var fields   = [],
        lines    = fs.readFileSync(dbCSVPath).toString().split("\n"),
        linesLen = lines.length,
        linesCB  = function(val, key) { fields[key] = val.replace(/"/g, ''); };

    for(var i = 0; i < linesLen; i++) {
      if(lines[i].indexOf('#') === 0) continue; // Check the line
      fields = [];
      lines[i].split(',').forEach(linesCB);
      if(fields.length !== 7) continue;         // Check the length
      dataSet.push(fields);                     // push the fields
    }

    return true;
  };

  // Returns the country information of the given IP address(es).
  ipSearch = function ipSearch(ip) {

    var result    = {data: {}, warnings: [], timeE: null},
        tsS       = new Date().getTime(),
        ipList    = {},
        ipListLen = 0,
        i         = 0;

    // Check the database
    if(!dbCSVCheck()) {
      return {error: 'Missing database file.'};
    } else if(!dbLoad()) {
      return {error: 'Database is not loaded!'};
    }

    // Find IP addresses
    var isAnyIP = false;

    if(ip) {
      if(typeof ip === 'string') ip = [ip];
      if(ip instanceof Array) {
        var ipF = null;
        for(i = 0; i < ip.length; i++) {
          ipF = ('' + ip[i]).trim();
          if(ipF) {
            var ts = ipF.split('.');

            if(ts.length !== 4) {
              result.warnings.push('Invalid IP address: ' + ipF);
            } else {
              ipList[ipF] = {
                ip: ipF,
                ipNum: parseInt(ts[3], 10) + (parseInt(ts[2], 10) * 256) + (parseInt(ts[1], 10) * 256 * 256) + (parseInt(ts[0], 10) * 256 * 256 * 256),
                registery: null,
                assigned: null,
                coCode2: null,
                coCode3: null,
                country: null,
                time: null,
                found: false
              };
              ipListLen++;
              isAnyIP = true;
            }
          }
        }
      }
    }
    if(!isAnyIP) {
      return {error: 'Missing IP address.'};
    } else if(ipListLen > 10) {
      return {error: 'Too many IP addresses. (' + ipListLen + '/10)'};
    }

    var dsLen    = dataSet.length,
        fields   = [],
        doneCntr = 0;

    for(i = 0; i < dsLen; i++) {
      fields = dataSet[i];

      for(var key in ipList) {
        if(ipList[key].found === false && ipList[key].ipNum >= fields[0] && ipList[key].ipNum <= fields[1]) {
          var asgndDate         = new Date(fields[3]*1000);

          ipList[key].registery = fields[2];
          ipList[key].assigned  = fields[3];
          ipList[key].time      = new Date(asgndDate.toISOString().replace("Z", "+0" + (asgndDate.getTimezoneOffset()/60) + ":00")).toISOString().replace("T", " ").substr(0, 19);
          ipList[key].coCode2   = fields[4];
          ipList[key].coCode3   = fields[5];
          ipList[key].country   = fields[6];
          ipList[key].found     = true;
          doneCntr++;
        }
      }

      if(doneCntr === ipListLen) break;
    }

    result.data  = ipList;
    result.timeE = ((new Date().getTime())-tsS);

    return result;
  };

  // Listens http requests.
  listenHTTP = function listenHTTP(options) {

    var hostname = (options && options.hostname) ? options.hostname : 'localhost',
        port     = (options && options.port)     ? options.port     : 12080,
        resHdr   = {'Content-Type': 'application/json'};

    // create http server
    var server = http.createServer(function(req, res) {
      var up = url.parse(req.url, true, false);
      if(up && up.pathname) {
        var pathAry = up.pathname.split('/');
        if(pathAry[1] === 'search') {
          var ipAry = (up.query && up.query.ip) ? ('' + up.query.ip).split(',') : null,
              ipS   = ipSearch(ipAry)
          ;
          if(!ipS.error) {
            res.writeHead(200, resHdr);
            res.end(JSON.stringify(ipS));
          } else {
            res.writeHead(404, resHdr);
            res.end(JSON.stringify({code: '404', message: ipS.error}));
          }
        } else if(pathAry[1]) {
          res.writeHead(501, resHdr);
          res.end(JSON.stringify({code: '501', message: 'Not Implemented'}));
        } else {
          res.writeHead(200, resHdr);
          res.end();
        }
      }
    }).listen(port, hostname, function() {
      console.log('Server is listening on ' + server.address().address + ':' + server.address().port);
    });
  };

  return {
    dataDirCheck: dataDirCheck,
    dbFileCheck: dbFileCheck,
    dbCSVCheck: dbCSVCheck,
    dbCSVCheckExp: dbCSVCheckExp,
    dbGet: dbGet,
    dbLoad: dbLoad,
    ipSearch: ipSearch,
    listenHTTP: listenHTTP
  };
}();